### PR TITLE
Fix race condition in notifier/Manager.sendAll() function

### PR DESCRIPTION
**Pull request by liguozhong**
Fix race condition in notifier/Manager.sendAll() function.
image

### CODE REVIEW

**Comment by roidelapluie** [comment]
If I am right the cancel should be in the goroutine, too

**Comment by liguozhong** [comment]
> If I am right the cancel should be in the goroutine, too

done,thanks.

image

**Comment by roidelapluie** [comment]
I think we can create the with timeout in the goroutine so we don't need to pass cancel() ?

**Comment by liguozhong** [comment]
> I think we can create the with timeout in the goroutine so we don't need to pass cancel() ?

done,I agree with that
image

File: notifier/notifier.go [hunk 1]
```diff
@@ -530,7 +530,7 @@ func (n *Manager) sendAll(alerts ...*Alert) bool {
 				n.metrics.sent.WithLabelValues(url).Add(float64(len(alerts)))
 
 				wg.Done()
-			}(ams.client, am.url().String())
+			}(ctx, cancel, ams.client, am.url().String())
```
Comment (roidelapluie) [inline notifier/notifier.go]: I do not think there is a race with ctx and cancel. Why do you think so? They are scoped to the inner loop.

**Comment by roidelapluie** [comment]
So this PR should only move cancel() in the go routine.

**Comment by liguozhong** [comment]
> So this PR should only move cancel() in the go routine.

yes,done.

**Comment by liguozhong** [comment]
> So this PR should only move cancel() in the go routine.
> 
> yes,done.

image

File: notifier/notifier.go (hunk 1 above)
Comment (codesome) [reply notifier/notifier.go]: ^ same. Even in https://github.com/prometheus/prometheus/pull/6907

**Pull request closed without merging**